File: 0015-3sum/0015-3sum.js
var threeSum = function (nums) {
  let n = nums.length;

  nums.sort((a, b) => a - b);

  let res = [];

  for (let i = 0; i < n - 2; i++) {
    let left = i + 1;
    let right = n - 1;

    if (i > 0 && nums[i] == nums[i - 1]) continue;

    while (left < right) {
      let sum = nums[left] + nums[right] + nums[i];

      if (sum === 0) {
        res.push([nums[i], nums[left], nums[right]]);
        while (left < right && nums[left] == nums[left + 1]) left++;
        while (left < right && nums[right] == nums[right - 1]) right--;
        left++;
        right++;
      } else if (sum < 0) {
        left++;
      } else {
        right--;
      }
    }
  }

  return res;
};

File: 0015-3sum/0015-3sum.py
class Solution:
    def threeSum(self, nums):
        nums.sort()
        result = set()
        n = len(nums)

        i = 0

        while i < n - 2:
            left = i + 1
            right = n - 1

            while left < right:
                total = nums[i] + nums[left] + nums[right]

                if total < 0:
                    left += 1

                elif total > 0:
                    right -= 1

                else:
                    result.add((nums[i], nums[left], nums[right]))
                    left += 1
                    right -= 1

            i += 1

        return [list(triplet) for triplet in result]